### DISCUSSION TITLE
Compilation with Intel Compilers

### DISCUSSION THREAD

**Discussion by makeclean** [idaholab/moose#25547] [Q&A General]
Hi All

We've got access to some shiny new CPUs, however the devops people love them some Intel compilers, so I've been having some fun with that. We manage to get most of the way through the MOOSE compile, however we hit an error when building/linking the final libraries, for example
```
/home/dc-davi4/rds/rds-ukaea-ap001/dc-davi4/moose-dev-intel/moose/modules/navier_stokes/build/header_symlinks/NSFVBase.h(3480): error: no operator "++" matches these operands
            operand types are: ++ __gnu_cxx::__normal_iterator<std::__cxx11::string *, std::vector<std::__cxx11::string, std::allocator<std::__cxx11::string>>>
      for (const auto & param_name : parameter_names)
                                     ^
          detected during:
            instantiation of "void NSFVBase<BaseType>::checkBlockwiseConsistency<T>(std::__cxx11::string, std::vector<std::__cxx11::string, std::allocator<std::__cxx11::string>>) [with BaseType=Action, T=std::vector<std::__cxx11::string, std::allocator<std::__cxx11::string>>]" at line 3374
            instantiation of "void NSFVBase<BaseType>::checkFrictionParameterErrors() [with BaseType=Action]" at line 1113  
            instantiation of "void NSFVBase<BaseType>::addNSKernels() [with BaseType=Action]" at line 58 of "/home/dc-davi4/rds/rds-ukaea-ap001/dc-davi4/moose-dev-intel/moose/modules/navier_stokes/src/actions/NSFVAction.C"

/home/dc-davi4/rds/rds-ukaea-ap001/dc-davi4/moose-dev-intel/moose/modules/navier_stokes/build/header_symlinks/NSFVBase.h(3480): error: no operator "++" matches these operands
```
Just wondering if others have had fun recently building MOOSE with ICPC. The framework build is fine, but when I then proceed to build the modules we get this.

**Comment by makeclean** [comment]
The offending line if it helps 

``` 
for (const auto & param_name : parameter_names)
```

**Reply by makeclean** [reply]
not sure it helps, must be heavily templated

**Reply by lindsayad** [reply]
If Intel compilers can't do a range-based for loop iteration over an STL container, that is embarrassing

**Reply by lindsayad** [reply]
@makeclean if your devops people are insistent on using Intel compilers then I would suggest looking at this thread starting at about https://github.com/idaholab/moose/issues/12713#issuecomment-455632711. Perhaps they can contact Intel Premier Support.

**Reply by permcody** [reply]
If it's just a line or two like that, feel free to drop back to older syntax to see if you can get it working. This is exactly why we dropped support though in the first place. Intel is just, behind...

**Reply by lindsayad** [reply]
range-based for has been around since the 2011 standard, so we just shouldn't be dealing with this kind of thing in 2023

**Comment by GiudGiud** [comment]
Pretty stumped why an iterator would not have ++ defined. Let s see what I can find on google

**Reply by GiudGiud** [reply]
no luck. Which version of the intel compilers is this? I could try on INL clusters

This isnt the reason we do not support intel compilers. I think there's some recurring threading issues with them.
@permcody or @milljm will know why

**Reply by GiudGiud** [reply]
@lindsayad will know too

**Reply by permcody** [reply]
We dropped support after a lot of the newer syntax introduced by C++11 and 14 just didn't work. MOOSE was starting to get sprinkled full of `#ifdefs` with the new sleek syntax and the the same block of code written in the older C++03 standards to work on Intel. It was infuriating and starting to become a real issue as we had to routinely spend time with CIVET machines in an attempt to replicate these issues. The team pushed hard to just drop it. Alex linked the right issue where Intel tried to ask us what was wrong, but that didn't really go anywhere. It's like Intel just isn't trying to keep up with standards.
